### ISSUE TITLE
AsyncCqlTemplate Cached Prepared Statements [DATACASS-591]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-cassandra#761]
**[Michael Bazos](https://jira.spring.io/secure/ViewProfile.jspa?name=mbazos)** opened **[DATACASS-591](https://jira.spring.io/browse/DATACASS-591?redirect=false)** and commented

I upgraded from spring data cassandra 1.5.x to 2.0.9 and noticed when using the AsyncCqlTemplate that there is no default implementations to cache prepared statements like there is for CqlTemplate.

 

Without the prepared statements I get a lot of warnings from cassandra as follows:

```java
WARN com.datastax.driver.core.Cluster - Re-preparing already prepared query is generally an anti-pattern and will likely affect performance. Consider preparing the statement only once. Query='...`
```

Even though this is asynchronous operation I think it makes sense to provide a cached prepared statement. I was able to fix this by doing the following...which is similar in nature to `org.springframework.data.cassandra.core.cql.support.MapPreparedStatementCache`

 

If you believe this is the correct approach I can clean this up and put a PR in. Please let me know.

```java
public class AsyncCachedPreparedStatementCreator implements AsyncPreparedStatementCreator {

    private final String cql;

    private final Map<CacheKey, com.google.common.util.concurrent.ListenableFuture<PreparedStatement>> cache;

    AsyncCachedPreparedStatementCreator(String cql,
                                        Map<CacheKey, com.google.common.util.concurrent.ListenableFuture<PreparedStatement>> cache) {
        Assert.notNull(cache, "cql must not be null");
        Assert.notNull(cache, "Cache must not be null");
        
        this.cql = cql;
        this.cache = cache;
    }

    @Override
    public ListenableFuture<PreparedStatement> createPreparedStatement(Session session) throws DriverException {
        CacheKey cacheKey = new CacheKey(session, cql);

        com.google.common.util.concurrent.ListenableFuture<PreparedStatement> future = cache.computeIfAbsent(cacheKey,
                                                                                                             k -> session.prepareAsync(cql));
        return new GuavaListenableFutureAdapter<>(future, new CassandraExceptionTranslator());
    }

    private class CacheKey {

        final Cluster cluster;
        final String  keyspace;
        final String  cql;

        CacheKey(Session session, String cql) {

            this.cluster = session.getCluster();
            this.keyspace = session.getLoggedKeyspace();
            this.cql = cql;
        }

        @Override
        public boolean equals(final Object obj) {
            if (!(obj instanceof CacheKey)) {
                return false;
            }
            if (obj == this) {
                return true;
            }

            CacheKey rhs = (CacheKey) obj;
            return new EqualsBuilder()
                .append(cluster, rhs.cluster)
                .append(keyspace, rhs.keyspace)
                .append(cql, rhs.cql)
                .isEquals();
        }

        @Override
        public int hashCode() {
            return new HashCodeBuilder()
                .append(cluster)
                .append(keyspace)
                .append(cql)
                .toHashCode();
        }
    }
} 
```

---

**Affects:** 2.0.9 (Kay SR9)

**Issue Links:**
- [DATACASS-510](https://jira.spring.io/browse/DATACASS-510) Use Prepared Statements for CassandraTemplate and repository queries

- [DATACASS-833](https://jira.spring.io/browse/DATACASS-833) Deprecate cached PreparedStatementCreator support classes
 (_**"is superseded by"**_)
